tensorflow/compiler/tf2tensorrt/trt_convert.cc

```diff
@@ -0,0 +1,389 @@
+  bool profile_generation_mode = params.use_dynamic_shape && inputs.size() > 1;
+  if (profile_generation_mode) {
+    TF_RETURN_IF_ERROR(ToggleProfileGenerationMode(&segmented_graph_def, true));
+  }
+  TF_RETURN_IF_ERROR(session->Create(segmented_graph_def));
+  for (auto const& input : inputs) {
+    TF_RETURN_IF_ERROR(RunHelper(session, signature, input));
+  }
+  string prefix = "";
+  if (profile_generation_mode) {
+    prefix = "TrtBuildStep";
+    TF_RETURN_IF_ERROR(
+        ToggleProfileGenerationMode(&segmented_graph_def, false));
+    VLOG(3) << "Importing graph with _profile_generation_mode disabled";
+    TF_RETURN_IF_ERROR(
+        ImportGraphDefHelper(session, segmented_graph_def, prefix));
+  }
+  TF_RETURN_IF_ERROR(RunHelper(session, signature, *inputs.begin(), prefix));
+  return Status::OK();
+}
+
+// Returns the resource manager associated with the node.
+Status GetResourceManager(const NodeDef& node, Session* session,
+                          ResourceMgr** rm) {
+  const DeviceMgr* device_mgr;
+  TF_RETURN_IF_ERROR(session->LocalDeviceManager(&device_mgr));
+  Device* device;
+  string device_name = node.device().empty()
+                           ? "/job:localhost/replica:0/task:0/device:GPU:0"
+                           : node.device();
+  TF_RETURN_IF_ERROR(device_mgr->LookupDevice(device_name, &device));
+  *rm = device->resource_manager();
+  return Status::OK();
+}
+
+// Looks up the cache resurce associated with the TRT node.
+Status GetEngineCacheResource(const NodeDef& node, Session* session,
+                              TRTEngineCacheResource** resource) {
+  ResourceMgr* rm;
+  TF_RETURN_IF_ERROR(GetResourceManager(node, session, &rm));
+
+  absl::string_view resource_name = node.name();
+  size_t last_slash = resource_name.find_last_of('/');
+  if (last_slash != absl::string_view::npos) {
+    resource_name.remove_prefix(last_slash + 1);
+  }
+  const std::string container(kTfTrtContainerName);
+  *resource = nullptr;
+  TF_RETURN_IF_ERROR(
+      rm->Lookup(container, std::string(resource_name), resource));
+  if (resource == nullptr || (*resource)->cache_.size() == 0) {
+    return errors::Internal("Engine cache not found for", resource_name);
+  }
+  return Status::OK();
+}
+
+// Looks up the engine from the engine cache, and serializes the engine.
+Status ReadSerializedEngine(
+    const NodeDef& node, Session* session,
+    TrtUniquePtrType<nvinfer1::IHostMemory>* engine_data) {
+  TRTEngineCacheResource* resource;
+  TF_RETURN_IF_ERROR(GetEngineCacheResource(node, session, &resource));
+  core::ScopedUnref unref_cache_res(resource);
+  if (resource->cache_.size() > 1) {
+    return errors::Internal(
+        "Multiple engines found, but we can only serialize one");
+  }
+  const std::unique_ptr<EngineContext>& engine =
+      resource->cache_.begin()->second;
+  if (!engine || !engine->cuda_engine) {
+    // TODO(tfeher): Engine could have failed to build. Handle that correctly.
```

Comment:
What will the "TODO" be doing?
If the TRTEngineOp is not executed, we can't build the engine. 
I think we may just issue a warning and skip the TRTEngineOp?